# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

In [1]:
import pandas as pd
import numpy as np


In [2]:
horses = pd.read_csv('horse_data.csv', sep = ',', na_values = '?', names = ['Surgery', 
                                                'Age', 
                                                'Hospital_number', 
                                                'rectal_temperature', 
                                                'Puls', 
                                                'respiratory_rate', 
                                                'Exriment_temp', 
                                                'peripheral_pulse',
                                               'mucous_membranes',
                                               'capillary_refill_time',
                                               'pain',
                                               'peristalsis',
                                               'abdominal_distension',
                                               'nasogastric_tube',
                                               'nasogastric_reflux',
                                               'nasogastric_reflux PH',
                                               'rectal_examination',
                                               'abdomen',
                                               'packed_cell_volume',
                                               'total_protein',
                                               'abdominocentesis_appearance',
                                               'abdomcentesis_total_protein',
                                               'outcome',
                                               'urgical_lesion',
                                               'type_of_lesion 1',
                                               'type_of_lesion 2',
                                               'type_of_lesion 3',
                                               'cp_data'])
horses.head()

Surgery  Age  Hospital_number  rectal_temperature   Puls  respiratory_rate  \
0      2.0    1           530101                38.5   66.0              28.0   
1      1.0    1           534817                39.2   88.0              20.0   
2      2.0    1           530334                38.3   40.0              24.0   
3      1.0    9          5290409                39.1  164.0              84.0   
4      2.0    1           530255                37.3  104.0              35.0   

   Exriment_temp  peripheral_pulse  mucous_membranes  capillary_refill_time  \
0            3.0               3.0               NaN                    2.0   
1            NaN               NaN               4.0                    1.0   
2            1.0               1.0               3.0                    1.0   
3            4.0               1.0               6.0                    2.0   
4            NaN               NaN               6.0                    2.0   

   ...  packed_cell_volume  total_protein  abdominocentesis_appearance  \
0  ...                45.0            8.4                          NaN   
1  ...                50.0           85.0                          2.0   
2  ...                33.0            6.7                          NaN   
3  ...                48.0            7.2                          3.0   
4  ...                74.0            7.4                          NaN   

   abdomcentesis_total_protein  outcome  urgical_lesion  type_of_lesion 1  \
0                          NaN      2.0               2             11300   
1                          2.0      3.0               2              2208   
2                          NaN      1.0               2                 0   
3                          5.3      2.0               1              2208   
4                          NaN      2.0               2              4300   

   type_of_lesion 2  type_of_lesion 3  cp_data  
0                 0                 0        2  
1                 0                 0        2  
2                 0                 0        1  
3                 0                 0        1  
4                 0                 0        2  

[5 rows x 28 columns]

In [3]:
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Surgery                      299 non-null    float64
 1   Age                          300 non-null    int64  
 2   Hospital_number              300 non-null    int64  
 3   rectal_temperature           240 non-null    float64
 4   Puls                         276 non-null    float64
 5   respiratory_rate             242 non-null    float64
 6   Exriment_temp                244 non-null    float64
 7   peripheral_pulse             231 non-null    float64
 8   mucous_membranes             253 non-null    float64
 9   capillary_refill_time        268 non-null    float64
 10  pain                         245 non-null    float64
 11  peristalsis                  256 non-null    float64
 12  abdominal_distension         244 non-null    float64
 13  nasogastric_tube    

# Выбор столбцов и базовое изучение

In [4]:
horses_new = horses[['Surgery', 'Age', 'rectal_temperature', 'Puls', 'capillary_refill_time','abdominal_distension','total_protein','outcome']]
horses_new.head()

Surgery  Age  rectal_temperature   Puls  capillary_refill_time  \
0      2.0    1                38.5   66.0                    2.0   
1      1.0    1                39.2   88.0                    1.0   
2      2.0    1                38.3   40.0                    1.0   
3      1.0    9                39.1  164.0                    2.0   
4      2.0    1                37.3  104.0                    2.0   

   abdominal_distension  total_protein  outcome  
0                   4.0            8.4      2.0  
1                   2.0           85.0      3.0  
2                   1.0            6.7      1.0  
3                   4.0            7.2      2.0  
4                   NaN            7.4      2.0

In [5]:
for col in horses_new.columns:
    pct_missing = horses_new[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

Surgery - 0.3%
Age - 0.0%
rectal_temperature - 20.0%
Puls - 8.0%
capillary_refill_time - 10.7%
abdominal_distension - 18.7%
total_protein - 11.0%
outcome - 0.3%


In [6]:
print(horses_new.Surgery .mean())
print(horses_new.Surgery .median())
print(horses_new.Surgery .mode()[0])

print(horses_new.Surgery .std())
print(horses_new.Surgery .var())

1.3979933110367893
1.0
1.0
0.49030464432881815
0.24039864425040888


Среднее по больнице 1.4, причем мода и медиана совпадают и равны 1, что говорит нам о том, что в большенстве случаев лошади была сделана операция

In [7]:
print(horses_new.Age.mean())
print(horses_new.Age.median())
print(horses_new.Age.mode()[0])

print(horses_new.Age.std())
print(horses_new.Age.var())

1.64
1.0
1
2.1739719055576163
4.726153846153814


Возраст, мода и медеиана совпадают и равны 1, значит в большенстве у нас старые лошади.

In [8]:
print(horses_new.rectal_temperature.mean())
print(horses_new.rectal_temperature.median())
print(horses_new.rectal_temperature.mode()[0])

print(horses_new.rectal_temperature.std())
print(horses_new.rectal_temperature.var())

38.16791666666669
38.2
38.0
0.7322886641121578
0.5362466875871686


нормальная температура считается 37.8, все средние выше этого показателя, делаем вывод, что большая часть лошадей имеет повышенную ьемпературу тела

In [9]:
print(horses_new.Puls.mean())
print(horses_new.Puls.median())
print(horses_new.Puls.mode()[0])

print(horses_new.Puls.std())
print(horses_new.Puls.var())

71.91304347826087
64.0
48.0
28.630556660735003
819.7087747035575


30-40 нормальный пульс для взрослой лошади, иногда быывает 20-30, 
у нас средние показатели намного выше и также довольно больше стандартное отклонение

In [10]:
print(horses_new.capillary_refill_time.mean())
print(horses_new.capillary_refill_time.median())
print(horses_new.capillary_refill_time.mode()[0])

print(horses_new.capillary_refill_time.std())
print(horses_new.capillary_refill_time.var())

1.3059701492537314
1.0
1.0
0.47762853540401734
0.22812901783218667


клинический показатель времени заполения капиляров - средний равен 1, что дает нам время <=3х секундам

In [11]:
print(horses_new.abdominal_distension.mean())
print(horses_new.abdominal_distension.median())
print(horses_new.abdominal_distension.mode()[0])

print(horses_new.abdominal_distension.std())
print(horses_new.abdominal_distension.var())

2.2663934426229506
2.0
1.0
1.0651308894422629
1.1345038116440658


Средний показатель вдутия живота довольно сильно различается, мода равна 1, значит чаще вздутия нет

In [12]:
print(horses_new.total_protein.mean())
print(horses_new.total_protein.median())
print(horses_new.total_protein.mode()[0])

print(horses_new.total_protein.std())
print(horses_new.total_protein.var())

24.456928838951317
7.5
6.5
27.475009470785064
754.8761454197289


Нормальный уровень белка 6-7,5, у нас среднее в 4 раза больше, высокий уровень разброса

In [13]:
print(horses_new.outcome.mean())
print(horses_new.outcome.median())
print(horses_new.outcome.mode()[0])

print(horses_new.outcome.std())
print(horses_new.outcome.var())

1.5518394648829432
1.0
1.0
0.7371869448054933
0.5434445915916574


мода и медиана совпадают и равны 1, значит чаще лошади выживали

In [14]:
horses_new.describe().round(2)

Surgery     Age  rectal_temperature    Puls  capillary_refill_time  \
count   299.00  300.00              240.00  276.00                 268.00   
mean      1.40    1.64               38.17   71.91                   1.31   
std       0.49    2.17                0.73   28.63                   0.48   
min       1.00    1.00               35.40   30.00                   1.00   
25%       1.00    1.00               37.80   48.00                   1.00   
50%       1.00    1.00               38.20   64.00                   1.00   
75%       2.00    1.00               38.50   88.00                   2.00   
max       2.00    9.00               40.80  184.00                   3.00   

       abdominal_distension  total_protein  outcome  
count                244.00         267.00   299.00  
mean                   2.27          24.46     1.55  
std                    1.07          27.48     0.74  
min                    1.00           3.30     1.00  
25%                    1.00           6.50     1.00  
50%                    2.00           7.50     1.00  
75%                    3.00          57.00     2.00  
max                    4.00          89.00     3.00

# Выбросы

In [15]:
q1 = horses_new.rectal_temperature.quantile(0.25)
q3 = horses_new.rectal_temperature.quantile(0.75)
print(q1, q3)
iqr = q3 - q1
lower_bound = q1 - (2 * iqr) 
upper_bound = q3 + (2 * iqr)
remove_outliers = horses_new[~horses_new.rectal_temperature.between(lower_bound, upper_bound, inclusive=True)].sort_values('rectal_temperature')
remove_outliers

37.8 38.5


Surgery  Age  rectal_temperature   Puls  capillary_refill_time  \
44       1.0    1                35.4  140.0                    2.0   
141      2.0    1                36.0   42.0                    1.0   
238      2.0    1                36.1   88.0                    1.0   
281      2.0    1                40.0   78.0                    1.0   
54       2.0    1                40.3  114.0                    2.0   
..       ...  ...                 ...    ...                    ...   
274      1.0    1                 NaN   76.0                    NaN   
282      1.0    1                 NaN   70.0                    2.0   
288      1.0    1                 NaN    NaN                    NaN   
293      1.0    1                 NaN   78.0                    1.0   
295      1.0    1                 NaN  120.0                    2.0   

     abdominal_distension  total_protein  outcome  
44                    NaN           69.0      3.0  
141                   NaN            6.8      2.0  
238                   2.0            7.0      3.0  
281                   1.0            6.5      2.0  
54                    3.0            8.1      3.0  
..                    ...            ...      ...  
274                   4.0            NaN      3.0  
282                   2.0            7.5      2.0  
288                   NaN            NaN      1.0  
293                   NaN           62.0      3.0  
295                   NaN           65.0      3.0  

[67 rows x 8 columns]

In [16]:
q1 = horses_new.Puls.quantile(0.25)
q3 = horses_new.Puls.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - iqr 
upper_bound = q3 + (1.5 * iqr)
print(lower_bound, upper_bound)
remove_outliers = horses_new[~horses_new.Puls.between(lower_bound, upper_bound, inclusive=True)].sort_values('Puls')
remove_outliers

8.0 148.0


Surgery  Age  rectal_temperature   Puls  capillary_refill_time  \
41       2.0    9                39.0  150.0                    NaN   
275      1.0    9                38.8  150.0                    2.0   
55       1.0    9                38.6  160.0                    1.0   
3        1.0    9                39.1  164.0                    2.0   
255      1.0    9                38.8  184.0                    1.0   
5        2.0    1                 NaN    NaN                    1.0   
28       1.0    1                 NaN    NaN                    NaN   
52       2.0    1                 NaN    NaN                    1.0   
56       1.0    1                 NaN    NaN                    NaN   
58       1.0    1                 NaN    NaN                    NaN   
74       1.0    9                 NaN    NaN                    NaN   
78       1.0    1                 NaN    NaN                    3.0   
83       1.0    1                38.0    NaN                    2.0   
93       2.0    1                 NaN    NaN                    1.0   
115      2.0    1                 NaN    NaN                    1.0   
117      2.0    1                39.5    NaN                    2.0   
126      1.0    1                38.7    NaN                    1.0   
136      1.0    1                 NaN    NaN                    2.0   
151      1.0    1                39.3    NaN                    2.0   
159      2.0    1                 NaN    NaN                    1.0   
160      1.0    1                 NaN    NaN                    NaN   
174      2.0    1                 NaN    NaN                    NaN   
175      1.0    1                 NaN    NaN                    NaN   
182      1.0    1                38.0    NaN                    2.0   
204      1.0    1                39.2    NaN                    1.0   
207      1.0    1                 NaN    NaN                    NaN   
216      2.0    1                 NaN    NaN                    NaN   
227      1.0    9                 NaN    NaN                    NaN   
288      1.0    1                 NaN    NaN                    NaN   

     abdominal_distension  total_protein  outcome  
41                    NaN            8.5      1.0  
275                   2.0            6.2      2.0  
55                    4.0            NaN      2.0  
3                     4.0            7.2      2.0  
255                   3.0            3.3      2.0  
5                     2.0            NaN      1.0  
28                    NaN            NaN      2.0  
52                    1.0            7.7      1.0  
56                    NaN            6.7      1.0  
58                    3.0            5.9      2.0  
74                    NaN            4.9      2.0  
78                    3.0            5.9      2.0  
83                    4.0            7.8      2.0  
93                    3.0            6.5      2.0  
115                   2.0           70.0      1.0  
117                   3.0            6.7      3.0  
126                   2.0            6.5      1.0  
136                   3.0            5.9      2.0  
151                   2.0            NaN      2.0  
159                   2.0            7.5      1.0  
160                   NaN           58.0      1.0  
174                   4.0            NaN      2.0  
175                   NaN            NaN      1.0  
182                   4.0            7.8      2.0  
204                   2.0            6.6      1.0  
207                   NaN            8.7      2.0  
216                   NaN            NaN      1.0  
227                   NaN            6.5      1.0  
288                   NaN            NaN      1.0

In [17]:
q1 = horses_new.total_protein.quantile(0.25)
q3 = horses_new.total_protein.quantile(0.75)
print(q1, q3)
iqr = q3 - q1
lower_bound = q1 - 0.1*iqr
upper_bound = q3 + (2*iqr)
print(lower_bound, upper_bound)
remove_outliers = horses_new[~horses_new.total_protein.between(lower_bound, upper_bound, inclusive=True)].sort_values('total_protein', ascending = True)
remove_outliers

6.5 57.0
1.4499999999999993 158.0


Surgery  Age  rectal_temperature   Puls  capillary_refill_time  \
5        2.0    1                 NaN    NaN                    1.0   
17       2.0    1                37.5   48.0                    NaN   
25       2.0    1                37.8   60.0                    1.0   
28       1.0    1                 NaN    NaN                    NaN   
39       1.0    9                39.2  146.0                    NaN   
51       1.0    1                37.4   84.0                    2.0   
55       1.0    9                38.6  160.0                    1.0   
59       2.0    1                 NaN   96.0                    2.0   
72       1.0    1                37.7   56.0                    NaN   
81       1.0    1                37.3   40.0                    1.0   
101      1.0    1                 NaN   60.0                    2.0   
113      1.0    1                36.8   60.0                    NaN   
130      1.0    1                37.8   30.0                    NaN   
137      1.0    1                38.0   48.0                    1.0   
151      1.0    1                39.3    NaN                    2.0   
158      2.0    1                38.0   42.0                    1.0   
169      1.0    1                 NaN   65.0                    2.0   
174      2.0    1                 NaN    NaN                    NaN   
175      1.0    1                 NaN    NaN                    NaN   
177      1.0    1                 NaN   68.0                    1.0   
179      2.0    1                38.5  100.0                    2.0   
185      1.0    1                38.5  120.0                    2.0   
205      2.0    1                38.5  100.0                    2.0   
216      2.0    1                 NaN    NaN                    NaN   
218      1.0    1                38.1   40.0                    1.0   
220      1.0    1                39.2   88.0                    2.0   
230      1.0    1                38.5  104.0                    1.0   
243      1.0    1                38.4   92.0                    2.0   
245      2.0    1                 NaN   96.0                    2.0   
248      1.0    1                38.0  108.0                    1.0   
274      1.0    1                 NaN   76.0                    NaN   
288      1.0    1                 NaN    NaN                    NaN   
296      2.0    1                37.2   72.0                    2.0   

     abdominal_distension  total_protein  outcome  
5                     2.0            NaN      1.0  
17                    NaN            NaN      1.0  
25                    NaN            NaN      1.0  
28                    NaN            NaN      2.0  
39                    NaN            NaN      2.0  
51                    2.0            NaN      2.0  
55                    4.0            NaN      2.0  
59                    4.0            NaN      2.0  
72                    NaN            NaN      2.0  
81                    2.0            NaN      1.0  
101                   NaN            NaN      3.0  
113                   NaN            NaN      2.0  
130                   NaN            NaN      2.0  
137                   4.0            NaN      1.0  
151                   2.0            NaN      2.0  
158                   1.0            NaN      1.0  
169                   4.0            NaN      3.0  
174                   4.0            NaN      2.0  
175                   NaN            NaN      1.0  
177                   NaN            NaN      2.0  
179                   4.0            NaN      3.0  
185                   NaN            NaN      1.0  
205                   4.0            NaN      3.0  
216                   NaN            NaN      1.0  
218                   NaN            NaN      3.0  
220                   NaN            NaN      3.0  
230                   4.0            NaN      1.0  
243                   3.0            NaN      1.0  
245                   4.0            NaN      2.0  
248                   3.0     

Температура - есть несколько выбросов по границам темепратурного интервала, но они не критические, я считаю, что вполне может быть, что у лошади была температура 35 или 40 градусов, при нормально температуре тела 37,8, поэтому мы их оставляем

Пульс - присутвует несколько выбросов, пульс за пределом верхней границы, но не намного, причем у всех лошадей с таким пульсом была высокая температура, вероятно она была в шоковом состоянии, оставляем.

Уровень белка -если брать формулу для границ, как на занятии, то выбросов нет. Но мы видим, что имеем огромную разницу между первым и третим квантилем, и также большую меру разброса. Возможно при сильном обезвоживании он может достигиать высоких значений, оставляем.

# Замена пропусков

In [18]:
q1 = horses_new.Surgery.quantile(0.25)
q3 = horses_new.Surgery.quantile(0.75)
print(q1, q3)
iqr = q3 - q1
lower_bound = q1 - iqr
upper_bound = q3 + iqr
print(lower_bound, upper_bound)
remove_outliers = horses_new[~horses_new.Surgery.between(lower_bound, upper_bound, inclusive=True)].sort_values('Surgery', ascending = True)
remove_outliers

1.0 2.0
0.0 3.0


Surgery  Age  rectal_temperature  Puls  capillary_refill_time  \
132      NaN    1                38.0  48.0                    1.0   

     abdominal_distension  total_protein  outcome  
132                   1.0           73.0      NaN

In [19]:
q1 = horses_new.Age.quantile(0.25)
q3 = horses_new.Age.quantile(0.75)
print(q1, q3)
iqr = q3 - q1
lower_bound = q1 - iqr
upper_bound = q3 + iqr
print(lower_bound, upper_bound)
outliers = horses_new[~horses_new.Age.between(lower_bound, upper_bound, inclusive=True)].sort_values('Age', ascending = True)
outliers

1.0 1.0
1.0 1.0


Surgery  Age  rectal_temperature   Puls  capillary_refill_time  \
3        1.0    9                39.1  164.0                    2.0   
262      2.0    9                38.6   88.0                    NaN   
255      1.0    9                38.8  184.0                    1.0   
252      1.0    9                38.8  124.0                    1.0   
244      1.0    9                38.2  124.0                    1.0   
229      1.0    9                38.5  120.0                    NaN   
227      1.0    9                 NaN    NaN                    NaN   
212      1.0    9                38.9  120.0                    2.0   
191      2.0    9                 NaN  100.0                    1.0   
135      1.0    9                38.1  136.0                    1.0   
109      1.0    9                38.4   84.0                    1.0   
103      1.0    9                38.0  140.0                    1.0   
82       1.0    9                38.1  100.0                    1.0   
75       1.0    9                39.7  100.0                    2.0   
74       1.0    9                 NaN    NaN                    NaN   
55       1.0    9                38.6  160.0                    1.0   
41       2.0    9                39.0  150.0                    NaN   
39       1.0    9                39.2  146.0                    NaN   
23       1.0    9                38.3  130.0                    1.0   
16       1.0    9                 NaN  128.0                    2.0   
13       2.0    9                38.0   92.0                    1.0   
9        2.0    9                38.3   90.0                    1.0   
275      1.0    9                38.8  150.0                    2.0   
287      2.0    9                39.5   84.0                    1.0   

     abdominal_distension  total_protein  outcome  
3                     4.0            7.2      2.0  
262                   NaN            5.9      1.0  
255                   3.0            3.3      2.0  
252                   4.0            7.6      2.0  
244                   4.0            8.0      1.0  
229                   NaN           54.0      1.0  
227                   NaN            6.5      1.0  
212                   3.0            6.3      1.0  
191                   1.0            4.7      1.0  
135                   3.0            4.9      2.0  
109                   3.0            6.6      2.0  
103                   2.0            5.3      1.0  
82                    1.0            5.7      1.0  
75                    NaN           57.0      3.0  
74                    NaN            4.9      2.0  
55                    4.0            NaN      2.0  
41                    NaN            8.5      1.0  
39                    NaN            NaN      2.0  
23                    NaN           70.0      1.0  
16                    3.0            7.8      2.0  
13                    2.0            6.1      2.0  
9                     1.0            6.2      1.0  
275                   2.0            6.2      2.0  
287                   NaN            5.0      1.0

In [20]:
q1 = horses_new.capillary_refill_time.quantile(0.25)
q3 = horses_new.capillary_refill_time.quantile(0.75)
print(q1, q3)
iqr = q3 - q1
lower_bound = q1 - iqr
upper_bound = q3 + iqr
print(lower_bound, upper_bound)
outliers = horses_new[~horses_new.capillary_refill_time.between(lower_bound, upper_bound, inclusive=True)].sort_values('capillary_refill_time', ascending = True)
outliers

1.0 2.0
0.0 3.0


Surgery  Age  rectal_temperature   Puls  capillary_refill_time  \
17       2.0    1                37.5   48.0                    NaN   
19       2.0    1                39.4  110.0                    NaN   
22       1.0    1                38.6   42.0                    NaN   
28       1.0    1                 NaN    NaN                    NaN   
39       1.0    9                39.2  146.0                    NaN   
41       2.0    9                39.0  150.0                    NaN   
56       1.0    1                 NaN    NaN                    NaN   
58       1.0    1                 NaN    NaN                    NaN   
72       1.0    1                37.7   56.0                    NaN   
74       1.0    9                 NaN    NaN                    NaN   
90       2.0    1                38.0   52.0                    NaN   
113      1.0    1                36.8   60.0                    NaN   
114      1.0    1                38.4   44.0                    NaN   
130      1.0    1                37.8   30.0                    NaN   
149      2.0    1                38.7   48.0                    NaN   
160      1.0    1                 NaN    NaN                    NaN   
171      2.0    1                39.0   86.0                    NaN   
174      2.0    1                 NaN    NaN                    NaN   
175      1.0    1                 NaN    NaN                    NaN   
195      2.0    1                38.0   44.0                    NaN   
202      1.0    1                37.5   48.0                    NaN   
207      1.0    1                 NaN    NaN                    NaN   
216      2.0    1                 NaN    NaN                    NaN   
227      1.0    9                 NaN    NaN                    NaN   
229      1.0    9                38.5  120.0                    NaN   
232      1.0    1                38.5   30.0                    NaN   
262      2.0    9                38.6   88.0                    NaN   
274      1.0    1                 NaN   76.0                    NaN   
280      1.0    1                38.6   48.0                    NaN   
283      1.0    1                38.2   72.0                    NaN   
288      1.0    1                 NaN    NaN                    NaN   
299      1.0    1                37.2   40.0                    NaN   

     abdominal_distension  total_protein  outcome  
17                    NaN            NaN      1.0  
19                    3.0            8.7      1.0  
22                    1.0            7.2      1.0  
28                    NaN            NaN      2.0  
39                    NaN            NaN      2.0  
41                    NaN            8.5      1.0  
56                    NaN            6.7      1.0  
58                    3.0            5.9      2.0  
72                    NaN            NaN      2.0  
74                    NaN            4.9      2.0  
90                    NaN           86.0      1.0  
113                   NaN            NaN      2.0  
114                   3.0           77.0      1.0  
130                   NaN            NaN      2.0  
149                   NaN           63.0      1.0  
160                   NaN           58.0      1.0  
171                   3.0            5.8      2.0  
174                   4.0            NaN      2.0  
175                   NaN            NaN      1.0  
195                   2.0           64.0      1.0  
202                   NaN           55.0      3.0  
207                   NaN            8.7      2.0  
216                   NaN            NaN      1.0  
227                   NaN            6.5      1.0  
229                   NaN           54.0      1.0  
232                   NaN            7.7      1.0  
262                   NaN            5.9      1.0  
274                   4.0            NaN      3.0  
280                   NaN           67.0      1.0  
283                   NaN            6.4      1.0  
288                   NaN            NaN      1.0

In [31]:
q1 = horses_new.abdominal_distension.quantile(0.25)
q3 = horses_new.abdominal_distension.quantile(0.75)
print(q1, q3)
iqr = q3 - q1
lower_bound = q1 - 0.5*iqr
upper_bound = q3 + iqr
print(lower_bound, upper_bound)
outliers = horses_new[~horses_new.abdominal_distension.between(lower_bound, upper_bound, inclusive=True)].sort_values('abdominal_distension', ascending = True)
outliers

1.0 3.0
0.0 5.0


Surgery  Age  rectal_temperature   Puls  capillary_refill_time  \
4        2.0    1                37.3  104.0                    2.0   
17       2.0    1                37.5   48.0                    NaN   
23       1.0    9                38.3  130.0                    1.0   
25       2.0    1                37.8   60.0                    1.0   
28       1.0    1                 NaN    NaN                    NaN   
33       1.0    1                38.2   64.0                    1.0   
39       1.0    9                39.2  146.0                    NaN   
41       2.0    9                39.0  150.0                    NaN   
44       1.0    1                35.4  140.0                    2.0   
53       2.0    1                38.6   40.0                    1.0   
56       1.0    1                 NaN    NaN                    NaN   
60       2.0    1                37.8   48.0                    1.0   
62       1.0    1                37.8   88.0                    1.0   
68       2.0    1                 NaN   52.0                    1.0   
69       2.0    1                38.0   48.0                    1.0   
72       1.0    1                37.7   56.0                    NaN   
74       1.0    9                 NaN    NaN                    NaN   
75       1.0    9                39.7  100.0                    2.0   
87       2.0    1                 NaN   40.0                    1.0   
90       2.0    1                38.0   52.0                    NaN   
100      1.0    1                38.2   40.0                    1.0   
101      1.0    1                 NaN   60.0                    2.0   
102      2.0    1                38.3   40.0                    1.0   
106      1.0    1                38.3   52.0                    1.0   
113      1.0    1                36.8   60.0                    NaN   
130      1.0    1                37.8   30.0                    NaN   
134      1.0    1                 NaN  100.0                    2.0   
139      2.0    1                38.0   60.0                    1.0   
141      2.0    1                36.0   42.0                    1.0   
149      2.0    1                38.7   48.0                    NaN   
160      1.0    1                 NaN    NaN                    NaN   
161      1.0    1                38.3   42.0                    1.0   
175      1.0    1                 NaN    NaN                    NaN   
177      1.0    1                 NaN   68.0                    1.0   
185      1.0    1                38.5  120.0                    2.0   
202      1.0    1                37.5   48.0                    NaN   
207      1.0    1                 NaN    NaN                    NaN   
216      2.0    1                 NaN    NaN                    NaN   
217      1.0    1                38.0   76.0                    2.0   
218      1.0    1                38.1   40.0                    1.0   
220      1.0    1                39.2   88.0                    2.0   
227      1.0    9                 NaN    NaN                    NaN   
229      1.0    9                38.5  120.0                    NaN   
232      1.0    1                38.5   30.0                    NaN   
239      1.0    1                38.1   70.0                    1.0   
257      1.0    1                38.7   96.0                    1.0   
262      2.0    9                38.6   88.0                    NaN   
268      1.0    1                38.2   42.0                    1.0   
280      1.0    1                38.6   48.0                    NaN   
283      1.0    1                38.2   72.0                    NaN   
287      2.0    9                39.5   84.0                    1.0   
288      1.0    1                 NaN    NaN                    NaN   
290      2.0    1                38.6   45.0                    1.0   
293      1.0    1                 NaN   78.0                    1.0   
295      1.0    1                 NaN  120.0                    2.0   
299      1.0    1            

In [22]:
q1 = horses_new.outcome.quantile(0.25)
q3 = horses_new.outcome.quantile(0.75)
print(q1, q3)
iqr = q3 - q1
lower_bound = q1 - iqr
upper_bound = q3 + iqr
print(lower_bound, upper_bound)
outliers = horses_new[~horses_new.outcome.between(lower_bound, upper_bound, inclusive=True)].sort_values('outcome', ascending = True)
outliers

1.0 2.0
0.0 3.0


Surgery  Age  rectal_temperature  Puls  capillary_refill_time  \
132      NaN    1                38.0  48.0                    1.0   

     abdominal_distension  total_protein  outcome  
132                   1.0           73.0      NaN

Пропущенные значения в Surgery и outcome можно заменить модой, чаще всего встречающееся значение и это всего один случай в нашем дата сете.

Странное значение в графе возраст - 9 заменим на моду 1, тк в основном у нас были старые лошади, вероятно кто-то по ошибке ввел напрямую возраст лошади 9 лет, а не 1, 2.

rectal_temperature - ранее мы видели, что все средние у нас более менее одинаковы, поэто пропущенное значение заменим на среднее - 38.2.

Puls - показатель сильно скачет и не понятно от чего зависит, бывает, что у лошади с нормальной темипературой высокий пульс, а  у лошади  с низкой - высокий, так же нет зависимости от outcome. Поэтому предлагаю заменить пропущенные значения средним.

abdominal_distension - можно заметить, что если лошадь умерла или была усыплена (outcome = 3 или 2), то показатель вздутия живота у животного был 4 или 3, поэтому предлагаю заменить пропущенные значения на медиану с группировкой по outcome.

capillary_refill_time - так же как и у abdominal_distension при сметельных случаях насыщение кислородом было медленным, те показатель равен 2. Заменим им все пропущенные занчения на седиану с группировкой по outcome.

total_protein - не понятно на что заменять, значение скачет и не понятно от чего зависит. Опять же проедлагаю заменить на среднее.

И еще предлагаю удалить все строки, где больше 4х пропущенных значений. (Есть данныу в которых известно только была ли операция и что случилось, остальные данные пропущены, такие данные будем считать статистически не значимеми и можем их удалить)


In [36]:
horses_new.isna().sum(axis = 1).sort_values()

0      0
298    0
150    0
271    0
153    0
      ..
160    4
216    5
288    5
175    5
28     5
Length: 300, dtype: int64

In [79]:
horses_clean = horses_new.drop(horses_new[horses_new.isna().sum(axis = 1) > 4].index)


In [62]:
horses_clean.isna().sum(axis = 1).sort_values()
horses_clean.isna().sum()

Surgery                   1
Age                       0
rectal_temperature       56
Puls                     20
capillary_refill_time    28
abdominal_distension     52
total_protein            29
outcome                   1
dtype: int64

In [80]:
horses_clean.Surgery.fillna(horses_new.Surgery.mode()[0], inplace=True)

In [81]:
horses_clean.outcome.fillna(horses_new.outcome.mode()[0], inplace=True)

In [82]:
horses_clean.Age.fillna(horses_new.Age.mode()[0], inplace=True)

In [93]:
horses_clean.Puls.fillna(horses_new.Puls.mean(), inplace=True)

In [94]:
horses_clean.rectal_temperature.fillna(horses_new.rectal_temperature.mean(), inplace=True)

In [96]:
horses_clean.total_protein.fillna(horses_new.total_protein.mean(), inplace=True)

In [91]:
horses_clean.capillary_refill_time.fillna(horses_new.groupby(['outcome']).capillary_refill_time.transform('median'), inplace=True)

In [89]:
horses_clean.abdominal_distension.fillna(horses_new.groupby(['outcome']).abdominal_distension.transform('median'), inplace=True)

In [97]:
horses_clean.isna().sum()

Surgery                  0
Age                      0
rectal_temperature       0
Puls                     0
capillary_refill_time    0
abdominal_distension     0
total_protein            0
outcome                  0
dtype: int64